Source/DafnyCore/Compilers/CompilerErrors.cs

```diff
@@ -53,150 +53,333 @@ public enum ErrorId {
+
+Note that entry points that are intended as unit tests can be marked with `{:test}` instead.
 ");
 
     Add(ErrorId.c_Main_method_not_permitted,
     @"
+If dafny finds no methods marked `{:main}`, it then looks for a method named `Main`
+to use as the starting point for the program.
+This error occurs if the `Main` method that is found
+does not qualify as a main entry point because it violates one or more of the [rules](#sec-user-guide-main),
+as given by the reason in the error message.
 ");
 
     Add(ErrorId.c_function_has_no_body,
     @"
+The given function has no body, which is OK for verification, but not OK for compilation 
+--- the compiler does not know what content to give it.
+Even ghost functions must have bodies because body-less ghost functions are a form of unchecked assumptions.
 ");
 
     Add(ErrorId.c_test_function_must_be_compilable,
     @"
+Only compiled functions and methods may be tested using the `dafny test` command, 
+which tests everything marked with `{:test}`.
+However this function is a ghost function, and consequently cannot be tested during execution.
+If you want the function to be compiled,
+declare the function as a `function` instead of `ghost function` in Dafny 4 
+or as `function method` instead of `function` in Dafny 3.
 ");
 
     Add(ErrorId.c_invalid_synthesize_method,
     @"
+The `{:synthesize}` attribute is an experimental attribute used to create 
+a mock object for methods that do not have bodies.
+It is currently only available for compiling to C# and in conjunction with the Moq library.
+See the [reference manual section on {:synthesize}](../DafnyRef/DafnyRef#sec-synthesize-attr) for more detail.
 ");
 
     Add(ErrorId.c_method_has_no_body,
     @"
+To be part of a compiled program, each method must have a body.
+Ghost methods are the equivalent of unchecked assumptions
+so they too must have bodies.
 ");
 
     Add(ErrorId.c_assume_statement_may_not_be_compiled,
     @"
+A method may be parsed and verified even if an [assume statement](../DafnyRef/DafnyRef#sec-assume-statement) is present. 
+However, the assume statement is an explicit, unchecked assumption.
+Dafny does not allow programs with unchecked assumptions, that is, incompletely verified programs, to be compiled.
+The `{:axiom}` attribute can be used to tell Dafny that the assumption is to be considered an
+externally verified _axiom_, with the program author taking responsibility for its validity.
+
+If the assumption marked with `{:axiom}` is not actually valid, then the validity of the entire program is in question.
 ");
 
     Add(ErrorId.c_forall_statement_has_no_body,
     @"
+A method may be parsed and verified even if a [forall statement](../DafnyRef/DafnyRef#sec-forall-statement) is missing a body. 
+However, the body must be supplied before the program can be compiled,
+even if the method is `ghost`. Body-less foralls in ghost methods are 
+similar to unchecked assumptions.
```

Comment:
```suggestion
unchecked assumptions.
```

Is there any way in which they aren't exactly that?